restcomm/restcomm.http/src/main/java/org/restcomm/connect/http/CallsEndpoint.java

```diff
@@ -228,7 +236,82 @@ protected Response getCalls(final String accountSid, UriInfo info,
         String endTime = info.getQueryParameters().getFirst("EndTime");
         String parentCallSid = info.getQueryParameters().getFirst("ParentCallSid");
         String conferenceSid = info.getQueryParameters().getFirst("ConferenceSid");
-        String reverse = info.getQueryParameters().getFirst("Reverse");
+
+        // 'Reverse' URL parameter currently behaves a bit weird; it has no control over the direction of the ordering,
+        // as is the case for example in IncomingPhoneNumbersEndpoint. What it does is affecting paging instead,
+        // and the reason I think is that back when this was implemented to avoid too much work on BE we wanted Console
+        // to be able to get the right set of results, even if without ordering so that they are ordered in
+        // the client side.
+
+        // But now that sorting is implemented, I think we don't need that anymore and we should be ok ignoring it. Also,
+        // I believe having a boolean Reverse url parameter won't be able to scale well in the future if we introduce
+        // multiple SortBy fields for example.
+        //String reverse = info.getQueryParameters().getFirst("Reverse");
+
+        // Format for sorting URL paremeter is '?SortBy=<field>:<direction>', for example: '?SortBy=start_time:asc'. In the
+        // future we can extend it to multiple sort fields, like ?SortBy=<field1>:<direction1>,<field2>:<direction2>
+        String sortParameters = info.getQueryParameters().getFirst("SortBy");
+
+        CallDetailRecordFilter.Builder filterBuilder = CallDetailRecordFilter.Builder.builder();
+
+        String sortBy = null;
+        String sortDirection = null;
+
+        if (sortParameters != null && !sortParameters.isEmpty()) {
+            final String[] values = sortParameters.split(":", 2);
+            sortBy = values[0];
+            if (values.length > 1) {
+                sortDirection = values[1];
+                if (sortBy.isEmpty()) {
+                    return status(BAD_REQUEST).entity(buildErrorResponseBody("Error parsing the SortBy parameter: missing field to sort by", responseType)).build();
+                }
+                if (!sortDirection.equalsIgnoreCase("asc") && !sortDirection.equalsIgnoreCase("desc")) {
+                    return status(BAD_REQUEST).entity(buildErrorResponseBody("Error parsing the SortBy parameter: sort direction needs to be either \'asc\' or \'desc\'", responseType)).build();
+                }
+            }
+        }
+
+        ///////
+        if (sortBy != null) {
+            if (sortBy.equals("date_created")) {
```

Comment:
is "date_created" REST API parameter? i think is called "DateCreated" in REST API. the point of all this refactorign is to hide DB names at this layer...

Reply:
@jaimecasero I agree on that, the problem is that the names I'm using are taken from what the API is returning for the CDR entity in the REST response not the DB:

```
    "calls": [
        {
            "sid": "CA011c9e9b9f80441a92eb6de49b0b9645",
            "InstanceId": "IDbb080f9ed7c34ac78197b35c12b58ab4",
            "date_created": "Fri, 16 Sep 2016 13:48:14 +0000",
            "date_updated": "Fri, 16 Sep 2016 13:48:14 +0000",
            "account_sid": "ACd239d5cab74caaa43a594772e4724687",
            "to": "jean",
            "from": "antonis",
            "status": "update_forced",
            "price_unit": "USD",
            "direction": "Client-To-Client",
            "api_version": "2012-04-24",
            "caller_name": "antonis",
            "uri": "/2012-04-24/Accounts/ACd239d5cab74caaa43a594772e4724687/Calls/CL2735f4b1998c4c6ab8944451503e53ba.json",
            "subresource_uris": {
                "notifications": "/2012-04-24/Accounts/ACd239d5cab74caaa43a594772e4724687/Calls/CA011c9e9b9f80441a92eb6de49b0b9645/Notifications.json",
                "recordings": "/2012-04-24/Accounts/ACd239d5cab74caaa43a594772e4724687/Calls/CA011c9e9b9f80441a92eb6de49b0b9645/Recordings.json"
            }
        },
```

And I think we should be using the same names in our API to make it more consistent and intuitive.

So I believe the root cause here is that some of the keys in the response are the same as the db columns, right?

Reply:
Yeah @jaimecasero , something is wrong I'm guessing in the converter because the xml response is fine:
```
        <Call>
            <Sid>CA01...</Sid>
            <InstanceId>IDbb...</InstanceId>
            <DateCreated>Fri, 16 Sep 2016 13:48:14 +0000</DateCreated>
            <DateUpdated>Fri, 16 Sep 2016 13:48:14 +0000</DateUpdated>
            <ParentCallSid/>
            <AccountSid>ACd23...</AccountSid>
            <To>jean</To>
            <From>antonis</From>
            <PhoneNumberSid/>
            <Status>update_forced</Status>
            <StartTime/>
            <EndTime/>
            <Duration/>
            <Price>0.00</Price>
            <PriceUnit>USD</PriceUnit>
            <Direction>Client-To-Client</Direction>
            <AnsweredBy/>
            <ApiVersion>2012-04-24</ApiVersion>
            <ForwardedFrom/>
            <CallerName>antonis</CallerName>
            <Uri>/2012-04-24/Accounts/ACd239.../Calls/CL2...</Uri>
            <SubresourceUris>
                <Notifications>/2012-04-24/Accounts/ACd239...7/Calls/CA011c9.../Notifications</Notifications>
                <Recordings>/2012-04-24/Accounts/ACd.../Calls/CA011.../Recordings</Recordings>
            </SubresourceUris>
            <Ring__duration/>
        </Call>
```

Let me do some digging

Reply:
@jaimecasero, yep it's the response converter, check [here](https://github.com/RestComm/Restcomm-Connect/blob/4cb3186007f31a469d6a4738af7f671f9937137d/restcomm/restcomm.http/src/main/java/org/restcomm/connect/http/converter/AbstractConverter.java#L110), where is the json conversion, and right above is the xml which is fine. And this seems to be done generically and not just for CDRs. Any ideas why we are enforcing different property names? 

I was thinking that for the sorting fields we should use the documented attribute names (i.e. DateCreated, etc), which are the same with .xml, and I guess we can correct the .json return values when we come up with a new API version at some point. Wdyt?

Reply:
What I ended up doing here is introducing a bunch of constants in the CallsEndpoint, similar to what @jaimecasero did for the IncomingNumbers